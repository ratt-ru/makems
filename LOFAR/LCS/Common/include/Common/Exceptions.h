//#  Exceptions.h: Declaration and definition of SINFONI specific exceptions.
//#
//#  Copyright (C) 2002-2006
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: Exceptions.h 12371 2008-12-23 13:18:31Z loose $

#ifndef LOFAR_COMMON_EXCEPTIONS_H
#define LOFAR_COMMON_EXCEPTIONS_H

//# Includes
#include <Common/Exception.h>

namespace LOFAR
{
  //
  // This exception will be thrown when an assertion fails.
  //
  EXCEPTION_CLASS(AssertError,Exception);

  //
  // This exception will be thrown when an I/O error occurs.
  //
  EXCEPTION_CLASS(IOException, Exception);

  //
  // This exception will be thrown when a math error occurs.
  //
  EXCEPTION_CLASS(MathException, Exception);

  //
  // This exception will be thrown when a method is called that has not been
  // implemented (yet).
  //
  EXCEPTION_CLASS(NotImplemented, Exception);

  //
  // This exception will be thrown when an error occurs in the ParameterSet
  // classes.
  //
  EXCEPTION_CLASS(APSException, Exception);

} // namespace ASTRON

#endif
