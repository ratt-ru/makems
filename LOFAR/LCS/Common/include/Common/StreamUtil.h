//#  StreamUtil.h: useful stream manipulation methods.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: StreamUtil.h 9368 2006-10-03 14:02:18Z loose $

#ifndef LOFAR_COMMON_STREAMUTIL_H
#define LOFAR_COMMON_STREAMUTIL_H

// \file
// Useful stream manipulation methods.

//# Includes
#include <Common/LofarTypes.h>
#include <Common/lofar_string.h>
#include <Common/lofar_vector.h>
#include <Common/lofar_iosfwd.h>

namespace LOFAR
{
  // Handles indentation of text lines. Every time an Indent object is
  // constructed, the static member \c level is incremented by one, and every
  // time an Indent object is destructed \c level is decremented by one. To
  // increment the amount of indentation you simply create an Indent
  // object. When this object goes out of scope, the amount of indentation is
  // automagically decremented.
  class Indent {
  public:
    // Constructor. Increments indentation level.
    Indent() { lvl++; }
    // Destructor. Decrements indentation level.
    ~Indent() { lvl--; }
    // Return the amount of indentation.
    static uint level() { return lvl; }
    // Return the token to be printed per indentation level
    static const string& token() { return tok; }
  private:
    // Indentation level.
    static uint lvl;
    // Token to be printed per indentation level.
    static const string tok;
  };


  // Write a vector to an ostream with a given separator, prefix and postfix.
  template<class T>
  void writeVector (std::ostream& os, const std::vector<T>& vec,
		    const char* separator=",",
		    const char* prefix="[", const char* postfix="]")
  {
    os << prefix;
    for (uint i = 0; i < vec.size(); i++) {
      if (i > 0) os << separator;
      os << vec[i];
    }
    os << postfix;
  }


  // Print an indentation that depends on the number of Indent objects
  // currently in existence.
  inline ostream& indent(ostream& os) 
  {
    for (uint i = 0; i < Indent::level(); ++i) {
      os << Indent::token();
    }
    return os;
  }


  // Print the contents of a vector enclosed in square brackets, using a comma
  // as separator.
  // \note operator<<() must be defined for type \c T.
  template<typename T>
  inline ostream& operator<<(ostream& os, const vector<T>& v)
  {
    writeVector(os, v, ",", "[", "]");
    return os;
  }


} // namespace LOFAR

#endif
