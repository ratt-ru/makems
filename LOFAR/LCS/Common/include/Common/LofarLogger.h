//#  LofarLogger.h: Macro interface to the lofar logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: LofarLogger.h 11806 2008-09-28 21:07:32Z loose $

#ifndef LOFAR_COMMON_LOFARLOGGER_H
#define LOFAR_COMMON_LOFARLOGGER_H

// \file
// Macro interface to the lofar logging package.

//# Include the correct set of macros, depending on the availability of the
//# log4cplus or log4cxx packages.
#if defined(HAVE_LOG4CPLUS)
# include <Common/LofarLog4Cplus.h>
#elif defined(HAVE_LOG4CXX)
# include <Common/LofarLog4Cxx.h>
#else
# include <Common/LofarLogCout.h>
#endif

#include <Common/Exceptions.h>

namespace LOFAR 
{

// \ingroup Common
// \addtogroup LofarLogger
  // @{


//# --------------------- Common definitions ---------------
//
// \name Common definitions 
// @{

// The package name of the current package. This variable should be defined in
// the file lofar_config.h which is generated by the lofarconf script. Here,
// we will make sure that LOFARLOGGER_PACKAGE is always defined.
#ifndef LOFARLOGGER_PACKAGE
# define LOFARLOGGER_PACKAGE "Unknown_package"
#endif

// The full package name is defined as the concatenation of the package name
// and an optional sub-package name that may be defined by the user.
#ifdef LOFARLOGGER_SUBPACKAGE
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE "." LOFARLOGGER_SUBPACKAGE
#else
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE
#endif

// @}

//#------------------------ Assert and FailWhen -------------------------------
//#
//# NB: THROW is defined in package-dependant file.
//
// \name Assert and FailWhen
// \note The \c DBG version of \c ASSERT and \c FAILWHEN will only be in your
// compiled code when the (pre)compiler flag \c ENABLE_DBGASSERT is defined.
// @{

// If the condition of assert is NOT met a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERT(cond) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, "Assertion: " #cond); \
	} while(0)


// If the condition of assert is NOT met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERTSTR(cond,stream) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, \
			   "Assertion: " #cond "; " << stream); \
	} while(0)

// If the condition of failwhen is met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define FAILWHEN(cond) do { \
	if (cond) THROW(::LOFAR::AssertError, "Failtest: " #cond); \
	} while(0)

#ifdef ENABLE_DBGASSERT
#define DBGASSERT(cond)				ASSERT(cond)
#define DBGASSERTSTR(cond,stream)	ASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)			FAILWHEN(cond)
#else
#define DBGASSERT(cond)
#define DBGASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)
#endif

// @}
    
} // namespace LOFAR

#endif // file read before
