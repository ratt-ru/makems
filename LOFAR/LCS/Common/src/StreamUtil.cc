//#  StreamUtil.cc: implementation of the stream utilities.
//#
//#  Copyright (C) 2002-2005
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: StreamUtil.cc 9368 2006-10-03 14:02:18Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/StreamUtil.h>

namespace LOFAR
{
  // Initialize static variables.
  uint Indent::lvl = 0;
  const string Indent::tok = ". ";

} // namespace LOFAR
