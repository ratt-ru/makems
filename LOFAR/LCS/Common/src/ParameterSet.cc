//#  ParameterSet.cc: Implements a map of Key-Value pairs.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: ParameterSet.cc 12371 2008-12-23 13:18:31Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>


#include <Common/ParameterSet.h>
#include <Common/LofarLogger.h>
#include <Common/lofar_fstream.h>


namespace LOFAR {

//-------------------------- creation and destroy ---------------------------
static ParameterSet* globalParameterSetInstance = 0;

ParameterSet* globalParameterSet()
{
  if (globalParameterSetInstance == 0) {
    globalParameterSetInstance = new ParameterSet();
  }
  return (globalParameterSetInstance);
}

//-------------------------- creation and destroy ---------------------------
//
// Default constructor
//
ParameterSet::ParameterSet(KeyCompare::Mode	mode)
  : itsSet (new ParameterSetImpl(mode))
{}

ParameterSet::ParameterSet(bool caseInsensitive)
  : itsSet (new ParameterSetImpl(caseInsensitive ?
                                 KeyCompare::NOCASE : KeyCompare::NORMAL))
{}

//
// Construction by reading a parameter file.
//
ParameterSet::ParameterSet(const string& theFilename, bool caseInsensitive)
  : itsSet (new ParameterSetImpl(theFilename, caseInsensitive ?
                                 KeyCompare::NOCASE : KeyCompare::NORMAL))
{}

//
// Construction by reading a parameter file.
//
ParameterSet::ParameterSet(const string&	theFilename,
			   KeyCompare::Mode	mode)
  : itsSet (new ParameterSetImpl(theFilename, mode))
{}

ParameterSet::ParameterSet(const char*	theFilename,
			   KeyCompare::Mode	mode)
  : itsSet (new ParameterSetImpl(string(theFilename), mode))
{}

//
// Copying is allowed.
//
ParameterSet::ParameterSet(const ParameterSet& that)
  : itsSet (that.itsSet->incrCount())
{}

//
// operator= copying
//
ParameterSet& 
ParameterSet::operator=(const ParameterSet& that)
{
	if (this != &that) {
	  unlink();
	  itsSet = that.itsSet->incrCount();
	}
	return (*this);
}

//
//	Destructor
//
ParameterSet::~ParameterSet()
{
  unlink();
}

void ParameterSet::unlink()
{
  if (itsSet->decrCount() == 0) {
    delete itsSet;
  }
}

//
// operator<<
//
std::ostream&	operator<< (std::ostream& os, const ParameterSet &thePS)
{
	os << *thePS.itsSet;
	return os;
}

} // namespace LOFAR
