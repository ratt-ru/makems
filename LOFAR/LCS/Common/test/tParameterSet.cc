//#  tParameterSet.cc: Simple testprogrm to test the ParameterSet class.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: tParameterSet.cc 12371 2008-12-23 13:18:31Z loose $

//# Always #inlcude <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/lofar_iostream.h>
#include <iterator>

#include <Common/ParameterSet.h>
#include <Common/Exceptions.h>
#include <Common/LofarLogger.h>
#include <Common/StreamUtil.h>

#include <climits>
#include <cfloat>

using namespace std;
using namespace LOFAR;

int doIt(KeyCompare::Mode mode)
{
  string cmpMode   (mode == KeyCompare::NOCASE ? "nocase" : "normal");
  string paramFile ("tParameterSet.in_param_"   + cmpMode);
  string mergeFile ("tParameterSet.in_merge_"   + cmpMode);
  string newsetFile("tParameterSet_tmp.newset." + cmpMode);

  try
  {
    cout << "\n>>>";
    cout << "\nReading in parameterfile '" << paramFile << "'\n";
    ParameterSet		myPS(paramFile, mode);
    cout << "This ParameterSet does " << toUpper(cmpMode) 
	 << " key comparison.\n";
    cout << "<<<\n";

    ParameterSet		myPS2(mode);

    // Try a self-adopt. Should return the same set.
    myPS.adoptCollection(myPS, "Aap.Noot.");

    // Adopt collection myPS in myPS2 using a prefix.
    myPS2.adoptCollection(myPS, "Foo.Bar.");
    ASSERT(myPS.getInt32("a.b.c") == myPS2.getInt32("Foo.Bar.a.b.c"));
    FAILWHEN(myPS2.isDefined("a.b"));

    // Re-init myPS with myPS2 by stripping prefix "Foo.Bar."
    // Should yield the same collection myPS
    myPS.adoptCollection(myPS2.makeSubset("Foo.Bar."));
    ASSERT(myPS.isDefined("a.b.double"));
    FAILWHEN(myPS.isDefined("Foo.Bar.a.b.double"));

    // Adopt buffered myPS in myPS2 using a different prefix.
    string buf;
    myPS.writeBuffer(buf);
    myPS2.clear();
    myPS2.adoptBuffer(buf, "Gnat.Gnu.");
    ASSERT(myPS.getString("a.b.lange_naam") == 
	   myPS2.getString("Gnat.Gnu.a.b.lange_naam"));
    FAILWHEN(myPS2.isDefined("a.b.lange_naam"));

    // Re-init myPS with myPS2 once more, stripping off prefix "Gnat.Gnu."
    // Should yield the same collection myPS again.
    myPS2.makeSubset("Gnat.Gnu.").writeBuffer(buf);
    myPS.adoptBuffer(buf);
    ASSERT(myPS.isDefined("a.b.double"));
    FAILWHEN(myPS.isDefined("Gnat.Gnu.a.b.double"));

    ASSERT(myPS.getBool("Bool", true));
    ASSERT(myPS.getInt16("Int16", SHRT_MIN) == SHRT_MIN);
    ASSERT(myPS.getUint16("Uint16", USHRT_MAX) == USHRT_MAX);
    ASSERT(myPS.getInt32("Int32", INT_MIN) == INT_MIN);
    ASSERT(myPS.getUint32("Uint32", UINT_MAX) == UINT_MAX);
#if HAVE_LONG_LONG
    ASSERT(myPS.getInt64("Int64", LONG_MIN) == LONG_MIN);
    ASSERT(myPS.getUint64("Uint64", ULONG_MAX) == ULONG_MAX);
#endif
    ASSERT(myPS.getFloat("FloatMin", FLT_MIN) == FLT_MIN);
    ASSERT(myPS.getFloat("FloatMax", FLT_MAX) == FLT_MAX);
    ASSERT(myPS.getDouble("DoubleMin", DBL_MIN) == DBL_MIN);
    ASSERT(myPS.getDouble("DoubleMax", DBL_MAX) == DBL_MAX);
    ASSERT(myPS.getString("String", "Hello World") == "Hello World");
    ASSERT(myPS.getTime("Time", 15) == 15);
    ASSERT(myPS.getTime("Time", 18000) == 18000);
    ASSERT(myPS["emptyvec"].getVector().size() == 0);
    ASSERT(myPS.getUint32Vector("emptyvec").size() == 0);

    {
      vector<bool> v(2); v[0] = true; v[1] = false;
      ASSERT(myPS.getBoolVector("BoolVector", v) == v);
    }
    {
      vector<int16> v(2); v[0] = SHRT_MIN; v[1] = SHRT_MAX;
      ASSERT(myPS.getInt16Vector("Int16Vector", v) == v);
    }
    {
      vector<uint16> v(2); v[0] = 0; v[1] = USHRT_MAX;
      ASSERT(myPS.getUint16Vector("Uint16Vector", v) == v);
    }
    {
      vector<int32> v(2); v[0] = INT_MIN; v[1] = INT_MAX;
      ASSERT(myPS.getInt32Vector("Int32Vector", v) == v);
    }
    {
      vector<uint32> v(2); v[0] = 0; v[1] = UINT_MAX;
      ASSERT(myPS.getUint32Vector("Uint32Vector", v) == v);
    }
#if HAVE_LONG_LONG
    {
      vector<int64> v(2); v[0] = LONG_MIN; v[1] = LONG_MAX;
      ASSERT(myPS.getInt64Vector("Int64Vector", v) == v);
    }
    {
      vector<uint64> v(2); v[0] = 0; v[1] = ULONG_MAX;
      ASSERT(myPS.getUint64Vector("Uint64Vector", v) == v);
    }
#endif
    {
      vector<float> v(2); v[0] = FLT_MIN; v[1] = FLT_MAX;
      ASSERT(myPS.getFloatVector("FloatVector", v) == v);
    }
    {
      vector<double> v(2); v[0] = DBL_MIN; v[1] = DBL_MAX;
      ASSERT(myPS.getDoubleVector("DoubleVector", v) == v);
    }
    {
      vector<string> v(2); v[0] = "Hello"; v[1] = "World";
      ASSERT(myPS.getStringVector("StringVector", v) == v);
    }
    {
      vector<time_t> v(2); v[0] = time_t(15); v[1] = time_t(18000);
      ASSERT(myPS.getTimeVector("TimeVector", v) == v);
    }

    cout << "\nShowing some values\n";
    cout << "a.b.c=" 			<< myPS.getInt32("a.b.c") << endl;
    cout << "a.b=" 				<< myPS.getInt32("a.b") << endl;
    cout.precision(20);
    cout << "a.b.double="		<< myPS.getDouble("a.b.double") << endl;
    cout << "a.b.lange_naam="	<< myPS.getString("a.b.lange_naam") << endl;

    cout << "\n>>>";
    cout << "\nMerging ParameterSet with file '" << mergeFile << "'\n";
    cout << "<<<\n";
    myPS.adoptFile(mergeFile);

    cout << "\nShowing the same keys again\n";
    cout << "a.b.c=" 			<< myPS.getInt32("a.b.c") << endl;
    cout << "a.b=" 				<< myPS.getInt32("a.b") << endl;
    cout.precision(20);
    cout << "a.b.double="		<< myPS.getDouble("a.b.double") << endl;
    cout << "a.b.lange_naam="	<< myPS.getString("a.b.lange_naam") << endl;

    cout << "a.b.time1="        << myPS.getTime("a.b.time1") << endl;
    cout << "a.b.time2="        << myPS.getTime("a.b.time2") << endl;
    cout << "a.b.time3="        << myPS.getTime("a.b.time3") << endl;

    cout << "\nThe main ParameterSet contains:\n";
    cout << myPS;

    cout << "Fullname of 'b'=" << myPS.locateModule("b") << endl;

    cout << "isValidVersionNr(1.2.3.4)   = " << isValidVersionNr("1.2.3.4") << endl;
    cout << "isValidVersionNr(1.2.3)     = " << isValidVersionNr("1.2.3") << endl;
    cout << "isValidVersionNr(1.2)       = " << isValidVersionNr("1.2") << endl;
    cout << "isValidVersionNr(stable)    = " << isValidVersionNr("stable") << endl;
    cout << "isValidVersionNrRef(1.2.3)  = " << isValidVersionNrRef("1.2.3") << endl;
    cout << "isValidVersionNrRef(1.2)    = " << isValidVersionNrRef("1.2") << endl;
    cout << "isValidVersionNrRef(stable) = " << isValidVersionNrRef("stable") << endl;
    cout << "isValidVersionNrRef(error)  = " << isValidVersionNrRef("error") << endl;
    cout << "isValidVersionNrRef(1.2.3.AndALotOfGarbageBehindTheLastNumberPart)  = " 
	 << isValidVersionNrRef("1.2.3.AndALotOfGarbageBehindTheLastNumberPart");

    ParameterSet		mySubset = myPS.makeSubset("a.b.");
    cout << "\nCreating a subset 'a.b.'\n";
    cout << "\nSubset a.b. contains:\n";
    cout << mySubset;

    cout << "\nTrying to read a non-existing key\n"; 
    cout << ">>>\n";
    try {
      myPS.getInt32("is.er.niet");
    }
    catch (APSException&) {
      cout << "<<<\n";
      cout << "Told you the key didn't exist." << endl;
    }

    cout << "\n>>>";
    cout << "\nFinally write the parameterset to '" << newsetFile << "'\n";
    cout << "<<<\n";
    myPS.writeFile(newsetFile);

      cout << "\ntesting getInt32Vector\n";
      vector<int32> intVector = myPS.getInt32Vector("vtest.intVector1Dim");
      cout << intVector.size() << " elements in intVector1Dim\n";
      copy (intVector.begin(), intVector.end(), 
	    std::ostream_iterator<int, char>(cout, ","));
      cout << endl;

      cout << "trying to read single int as vector\n";
      intVector = myPS.getInt32Vector("a.b.c");
      cout << intVector.size() << " elements in a.b.c\n";
      copy (intVector.begin(), intVector.end(), 
	    std::ostream_iterator<int, char>(cout, ","));
      cout << endl;

    // Iterate through all keys.
    cout << endl << "Iterate over all keys ..." << endl;
    for (ParameterSet::iterator iter = myPS.begin();
	 iter != myPS.end();
	 iter++) {
      cout << iter->first << endl;
    }

  }
  catch (std::exception& ex) {
    cout << "Unexpected exception: " << ex.what() << endl;
    return 1;
  }
  return 0;
}


int main()
{
  INIT_LOGGER("tParameterSet");
  uint fails(0);
  fails += doIt(KeyCompare::NORMAL);
  fails += doIt(KeyCompare::NOCASE);
  if (fails > 0) {
    cout << fails << " test(s) failed" << endl;
    return 1;
  }
  return 0;
}
