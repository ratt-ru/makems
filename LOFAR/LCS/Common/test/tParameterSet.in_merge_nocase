# comment line
A.b.C=7
a.B.lange_naam="dit is een wat langere naam"

a.B.c     = 700			# outlining around = sign is possible.
A.c.a     = nog een afkorting
a.b.DOUBLE= 3.1415926535897932384626433832795	# this is just added to a.b.double
a.B.c.D.e.F.g.H='tjee wat een levels!'
A.B=0

A.b.time1=15 s
 a.B.time2=15m
a.b.TIME3=15 hour

VTest.IntVector1Dim = [ 3, 20485,
                     	002, 4904 , 25 ]
#vtest.intVector1Dim = [ 3, "20485, 002, '4904' ]
#vtest.intVector1Dim = [ 3, "20485, 002, '4904" ]

