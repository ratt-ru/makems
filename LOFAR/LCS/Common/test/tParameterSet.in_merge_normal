# comment line
a.b.c=7
a.b.lange_naam="dit is een wat langere naam"

a.b.c     = 700			# outlining around = sign is possible.
a.c.a     = nog een afkorting
a.b.double= 3.1415926535897932384626433832795# this is just added to a.b.double
a.b.c.d.e.f.g.h='tjee wat een levels!'
a.b=0

a.b.time1=15 s
 a.b.time2=15m
a.b.time3=15 hour

vtest.intVector1Dim = [ 3, 20485, 
                	002, 4904 , 25 ]
#vtest.intVector1Dim = [ 3, "20485, 002, '4904' ]
#vtest.intVector1Dim = [ 3, "20485, 002, '4904" ]

