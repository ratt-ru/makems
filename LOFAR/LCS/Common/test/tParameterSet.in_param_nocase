# comment line
A.B=7
# the next two lines overlap with the KVs in test.ps
a.b.lange_naam="dit is nu een andere naam geworden met extra spaties aan het einde  "
a.B.c=5
A.b.DOUBLE= 3.1415926	# we may put comment here

e.g=een voorbeeld
Egg=een ei
emptyvec=[]
