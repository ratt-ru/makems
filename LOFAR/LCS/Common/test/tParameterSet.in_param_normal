# comment line
a.b=7
# the next two lines overlap with the KVs in test.ps
a.b.lange_naam = "dit is nu een andere naam geworden met extra spaties "
                 "aan het einde  "   #and shows quoted continuation
a.b.c=5
a.b.double= 3.1415926	# we may put comment here

e.g=een voorbeeld
egg=een ei
emptyvec=[]
