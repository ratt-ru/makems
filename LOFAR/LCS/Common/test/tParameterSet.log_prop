# Property file to be use with the demo program testLogger.

# Configure the rootLogger
log4cplus.rootLogger=DEBUG, STDOUT
# Define the STDOUT appender
log4cplus.appender.STDOUT=log4cplus::ConsoleAppender
log4cplus.appender.STDOUT.layout=log4cplus::PatternLayout
# Do not use conversion pattern to get macine independent output.
log4cplus.appender.STDOUT.layout.ConversionPattern=%-5p [%x]%c{3} - %m%n
log4cplus.appender.STDOUT.filters.1=log4cplus::spi::StringMatchFilter
log4cplus.appender.STDOUT.filters.1.AcceptOnMatch=false
log4cplus.appender.STDOUT.filters.1.StringToMatch=data

# Define TRC at level INFO
log4cplus.logger.TRC=INFO




