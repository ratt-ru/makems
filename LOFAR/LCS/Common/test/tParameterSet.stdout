
>>>
Reading in parameterfile 'tParameterSet.in_param_normal'
This ParameterSet does NORMAL key comparison.
<<<

Showing some values
a.b.c=5
a.b=7
a.b.double=3.1415926000000000684
a.b.lange_naam=dit is nu een andere naam geworden met extra spaties aan het einde  

>>>
Merging ParameterSet with file 'tParameterSet.in_merge_normal'
<<<

Showing the same keys again
a.b.c=700
a.b=0
a.b.double=3.141592653589793116
a.b.lange_naam=dit is een wat langere naam
a.b.time1=15
a.b.time2=900
a.b.time3=54000

The main ParameterSet contains:
a.b=0
a.b.c=700
a.b.c.d.e.f.g.h='tjee wat een levels!'
a.b.double=3.1415926535897932384626433832795
a.b.lange_naam="dit is een wat langere naam"
a.b.time1=15 s
a.b.time2=15m
a.b.time3=15 hour
a.c.a=nog een afkorting
e.g=een voorbeeld
egg=een ei
emptyvec=[]
vtest.intVector1Dim=[ 3, 20485, 002, 4904 , 25 ]
Fullname of 'b'=a.
isValidVersionNr(1.2.3.4)   = 0
isValidVersionNr(1.2.3)     = 1
isValidVersionNr(1.2)       = 0
isValidVersionNr(stable)    = 0
isValidVersionNrRef(1.2.3)  = 1
isValidVersionNrRef(1.2)    = 0
isValidVersionNrRef(stable) = 1
isValidVersionNrRef(error)  = 0
isValidVersionNrRef(1.2.3.AndALotOfGarbageBehindTheLastNumberPart)  = 0
Creating a subset 'a.b.'

Subset a.b. contains:
c=700
c.d.e.f.g.h='tjee wat een levels!'
double=3.1415926535897932384626433832795
lange_naam="dit is een wat langere naam"
time1=15 s
time2=15m
time3=15 hour

Trying to read a non-existing key
>>>
<<<
Told you the key didn't exist.

>>>
Finally write the parameterset to 'tParameterSet_tmp.newset.normal'
<<<

testing getInt32Vector
5 elements in intVector1Dim
3,20485,2,4904,25,
trying to read single int as vector
1 elements in a.b.c
700,

Iterate over all keys ...
a.b
a.b.c
a.b.c.d.e.f.g.h
a.b.double
a.b.lange_naam
a.b.time1
a.b.time2
a.b.time3
a.c.a
e.g
egg
emptyvec
vtest.intVector1Dim

>>>
Reading in parameterfile 'tParameterSet.in_param_nocase'
This ParameterSet does NOCASE key comparison.
<<<

Showing some values
a.b.c=5
a.b=7
a.b.double=3.1415926000000000684
a.b.lange_naam=dit is nu een andere naam geworden met extra spaties aan het einde  

>>>
Merging ParameterSet with file 'tParameterSet.in_merge_nocase'
<<<

Showing the same keys again
a.b.c=700
a.b=0
a.b.double=3.141592653589793116
a.b.lange_naam=dit is een wat langere naam
a.b.time1=15
a.b.time2=900
a.b.time3=54000

The main ParameterSet contains:
A.B=0
a.B.c=700
a.B.c.D.e.F.g.H='tjee wat een levels!'
a.b.DOUBLE=3.1415926535897932384626433832795
a.B.lange_naam="dit is een wat langere naam"
A.b.time1=15 s
a.B.time2=15m
a.b.TIME3=15 hour
A.c.a=nog een afkorting
e.g=een voorbeeld
Egg=een ei
emptyvec=[]
VTest.IntVector1Dim=[ 3, 20485, 002, 4904 , 25 ]
Fullname of 'b'=a.
isValidVersionNr(1.2.3.4)   = 0
isValidVersionNr(1.2.3)     = 1
isValidVersionNr(1.2)       = 0
isValidVersionNr(stable)    = 0
isValidVersionNrRef(1.2.3)  = 1
isValidVersionNrRef(1.2)    = 0
isValidVersionNrRef(stable) = 1
isValidVersionNrRef(error)  = 0
isValidVersionNrRef(1.2.3.AndALotOfGarbageBehindTheLastNumberPart)  = 0
Creating a subset 'a.b.'

Subset a.b. contains:
c=700
c.D.e.F.g.H='tjee wat een levels!'
DOUBLE=3.1415926535897932384626433832795
lange_naam="dit is een wat langere naam"
time1=15 s
time2=15m
TIME3=15 hour

Trying to read a non-existing key
>>>
<<<
Told you the key didn't exist.

>>>
Finally write the parameterset to 'tParameterSet_tmp.newset.nocase'
<<<

testing getInt32Vector
5 elements in intVector1Dim
3,20485,2,4904,25,
trying to read single int as vector
1 elements in a.b.c
700,

Iterate over all keys ...
A.B
a.B.c
a.B.c.D.e.F.g.H
a.b.DOUBLE
a.B.lange_naam
A.b.time1
a.B.time2
a.b.TIME3
A.c.a
e.g
Egg
emptyvec
VTest.IntVector1Dim
